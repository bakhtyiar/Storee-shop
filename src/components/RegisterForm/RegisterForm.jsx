import React, {useContext} from 'react';
import {Button, Form, Modal} from "react-bootstrap";
import {Link, useNavigate} from "react-router-dom";
import {RootContext} from "../../contexts/root-context";
import * as yup from 'yup';
import {Formik} from 'formik';
import {registerUser} from "../../utils/methods";
import {routes} from "../../utils/constants";

const schema = yup.object().shape({
    username: yup.string()
        .min(6, 'Username is too short. Required minimum 6 symbols')
        .max(50, 'Username is too long. Required maximum 50 symbols')
        .required('Username is required'),
    email: yup.string()
        .email('Seems like wrong format of email')
        .required('Email is required'),
    password: yup.string()
        .min(6, 'Password is too short. Write at least 6 symbols')
        .required('Password is required'),
    terms: yup.bool().required().oneOf([true], 'Terms must be accepted'),
});

const RegisterForm = ({isHaveCloseButton = false}) => {
    const {authModalState: {onSwitchType, onHide}, authUserState: {onLogin}} = useContext(RootContext);
    const navigate = useNavigate();

    const handleSubmit = async (event) => {
        console.log('event', event);
        const {username, email, password} = event;
        const res = await registerUser(username, email, password);
        console.log('res', res);
        onLogin(res);
        onHide();
        navigate(`${routes.user.path}/${res.id}`);
    };

    return (
        <Formik
            validationSchema={schema}
            validateOnBlur
            onSubmit={handleSubmit}
            initialValues={{
                username: '',
                email: '',
                password: '',
                terms: false,
            }}
        >
            {({
                  handleSubmit,
                  handleChange,
                  handleBlur,
                  values,
                  touched,
                  errors,
              }) => (
                <Form noValidate onSubmit={handleSubmit}>
                    <Modal.Header closeButton={isHaveCloseButton}>
                        <Modal.Title id="contained-modal-title-vcenter">
                            Register
                        </Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        <Form.Group className="mb-3" controlId="formBasicUsername">
                            <Form.Label>Username</Form.Label>
                            <Form.Control
                                type="text"
                                placeholder="Enter username"
                                name="username"
                                value={values.username}
                                onChange={handleChange}
                                onBlur={handleBlur}
                                isValid={touched.username && !errors.username}
                                isInvalid={touched.username && !!errors.username}
                            />
                            <Form.Control.Feedback type="invalid">
                                {touched.username && errors.username}
                            </Form.Control.Feedback>
                        </Form.Group>

                        <Form.Group className="mb-3" controlId="formBasicEmail">
                            <Form.Label>Email</Form.Label>
                            <Form.Control
                                type="email"
                                placeholder="Enter email"
                                required
                                name="email"
                                value={values.email}
                                onChange={handleChange}
                                onBlur={handleBlur}
                                isValid={touched.email && !errors.email}
                                isInvalid={touched.email && !!errors.email}
                            />
                            <Form.Control.Feedback type="invalid">
                                {touched.email && errors.email}
                            </Form.Control.Feedback>
                            <Form.Text className="text-muted">
                                We'll never share your confidential data
                            </Form.Text>
                        </Form.Group>

                        <Form.Group className="mb-3" controlId="formBasicPassword">
                            <Form.Label>Password</Form.Label>
                            <Form.Control
                                type="password"
                                placeholder="Password"
                                required
                                name="password"
                                value={values.password}
                                onChange={handleChange}
                                onBlur={handleBlur}
                                isValid={touched.password && !errors.password}
                                isInvalid={touched.password && !!errors.password}
                            />
                            <Form.Control.Feedback type="invalid">
                                {touched.password && errors.password}
                            </Form.Control.Feedback>
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="formBasicCheckbox">
                            <Form.Check
                                type="checkbox"
                                label="Agree to terms and conditions"
                                required
                                feedback={errors.terms}
                                feedbackType="invalid"
                                name="terms"
                                onChange={handleChange}
                                onBlur={handleBlur}
                                isValid={touched.terms && !errors.terms}
                                isInvalid={touched.terms && !!errors.terms}
                            />
                        </Form.Group>
                    </Modal.Body>
                    <Modal.Footer className='d-flex justify-content-between'>
                        <Link onClick={onSwitchType}>Already have an account?</Link>
                        <Button variant="primary" type="submit">
                            Register
                        </Button>
                    </Modal.Footer>
                </Form>
            )}
        </Formik>
    );
};

export default RegisterForm;

