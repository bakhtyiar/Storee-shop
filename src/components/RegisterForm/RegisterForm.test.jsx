import {screen} from "@testing-library/react";
import {renderWithRouter} from "../../tests/helpers/renderWithRouter";
import RegisterForm from "./RegisterForm";
//todo: write interactive tests as for LoginForm component
describe("Basic tests", () => {
    test("Match snapshot", () => {
        renderWithRouter(<RegisterForm/>);
        expect(screen.getByTestId("register-form")).toMatchSnapshot();
    })
})