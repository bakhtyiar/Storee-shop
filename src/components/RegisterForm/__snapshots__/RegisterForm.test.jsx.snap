// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic tests Match snapshot 1`] = `
<form
  class=""
  data-testid="register-form"
  novalidate=""
>
  <div
    class="modal-header"
  >
    <div
      class="modal-title h4"
      id="contained-modal-title-vcenter"
    >
      Register
    </div>
  </div>
  <div
    class="modal-body"
  >
    <div
      class="mb-3"
    >
      <div
        class="col"
        data-testid="form-text-field"
      >
        <label
          class="form-label"
          for="formBasicUsername"
        >
          Username
        </label>
        <div
          class="input-group"
        >
          <input
            class="rounded form-control"
            id="formBasicUsername"
            name="username"
            placeholder="Enter username"
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
          />
        </div>
      </div>
    </div>
    <div
      class="mb-3"
    >
      <div
        class="col"
        data-testid="form-text-field"
      >
        <label
          class="form-label"
          for="formBasicEmail"
        >
          Email
        </label>
        <div
          class="input-group"
        >
          <input
            class="rounded form-control"
            id="formBasicEmail"
            name="email"
            placeholder="Enter email"
            type="email"
            value=""
          />
          <div
            class="invalid-feedback"
          />
        </div>
        <small
          class="text-muted form-text"
        >
          We'll never share your confidential data
        </small>
      </div>
    </div>
    <div
      class="mb-3"
    >
      <div
        class="col"
        data-testid="form-text-field"
      >
        <label
          class="form-label"
          for="formBasicPassword"
        >
          Password
        </label>
        <div
          class="input-group"
        >
          <input
            class="rounded form-control"
            id="formBasicPassword"
            name="password"
            placeholder="Enter password"
            type="password"
            value=""
          />
          <div
            class="invalid-feedback"
          />
        </div>
      </div>
    </div>
    <div
      class="mb-3"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="formBasicCheckbox"
          name="terms"
          required=""
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="formBasicCheckbox"
          title=""
        >
          Agree to terms and conditions
        </label>
      </div>
    </div>
  </div>
  <div
    class="d-flex justify-content-between modal-footer"
  >
    <a
      as="[object Object]"
      href="/"
    >
      Already have an account?
    </a>
    <button
      class="btn btn-primary"
      type="submit"
    >
      Register
    </button>
  </div>
</form>
`;
